File: linux/frontasks/store.py
# FronTasks — floating, always-on-top task list for Linux.
# Copyright (C) 2026 Ananias Filho
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

"""Modelo TaskItem + TaskStore, persistidos em JSON (mesmo schema do Mac,
exceto createdAt que aqui usa epoch Unix -- ver README seção 9)."""

import time
import uuid
from dataclasses import dataclass, field, asdict
from typing import List

import gi
gi.require_version("GObject", "2.0")
from gi.repository import GObject

from .config import tasks_path
from .persistence import atomic_write_json, safe_load_json


@dataclass
class TaskItem:
    id: str = field(default_factory=lambda: str(uuid.uuid4()))
    title: str = ""
    isDone: bool = False
    createdAt: float = field(default_factory=time.time)
    order: int = 0


def _validate_tasks(raw) -> List[TaskItem]:
    """Valida schema/tipos item a item. Um item individual inválido é
    descartado (não derruba a lista inteira); se `raw` não for nem uma
    lista, propaga a exceção pra safe_load_json tratar como corrupto."""
    if not isinstance(raw, list):
        raise ValueError("tasks.json: esperava uma lista")
    tasks = []
    for item in raw:
        if not isinstance(item, dict):
            continue
        try:
            tasks.append(
                TaskItem(
                    id=str(item["id"]) if "id" in item else str(uuid.uuid4()),
                    title=str(item.get("title", "")),
                    isDone=bool(item.get("isDone", False)),
                    createdAt=float(item.get("createdAt", time.time())),
                    order=int(item.get("order", 0)),
                )
            )
        except (TypeError, ValueError):
            continue
    return tasks


class TaskStore(GObject.Object):
    __gsignals__ = {
        "changed": (GObject.SignalFlags.RUN_FIRST, None, ()),
    }

    def __init__(self):
        super().__init__()
        self.tasks: List[TaskItem] = []
        self._path = tasks_path()
        self.load()

    # --- Persistência -----------------------------------------------

    def load(self):
        tasks = safe_load_json(self._path, validate=_validate_tasks)
        if tasks is None:
            self.tasks = []
            return
        self.tasks = tasks
        self.tasks.sort(key=lambda t: (t.order, t.createdAt))

    def save(self):
        data = [asdict(t) for t in self.tasks]
        atomic_write_json(self._path, data)

    # --- CRUD ---------------------------------------------------------

    def add(self, title: str):
        title = title.strip()
        if not title:
            return
        max_order = max((t.order for t in self.tasks), default=0)
        self.tasks.append(TaskItem(title=title, order=max_order + 1))
        self.save()
        self.emit("changed")

    def toggle(self, task_id: str) -> bool:
        for t in self.tasks:
            if t.id == task_id:
                t.isDone = not t.isDone
                self.save()
                self.emit("changed")
                return True
        return False

    def update_title(self, task_id: str, title: str) -> bool:
        title = title.strip()
        for i, t in enumerate(self.tasks):
            if t.id == task_id:
                if not title:
                    del self.tasks[i]
                else:
                    t.title = title
                self.save()
                self.emit("changed")
                return True
        return False

    def delete(self, task_id: str) -> bool:
        before = len(self.tasks)
        self.tasks = [t for t in self.tasks if t.id != task_id]
        if len(self.tasks) == before:
            return False
        self.save()
        self.emit("changed")
        return True

    def move(self, old_index: int, new_index: int) -> bool:
        if not (0 <= old_index < len(self.tasks)) or not (0 <= new_index < len(self.tasks)):
            return False
        if old_index == new_index:
            return False
        task = self.tasks.pop(old_index)
        self.tasks.insert(new_index, task)
        for i, t in enumerate(self.tasks):
            t.order = i
        self.save()
        self.emit("changed")
        return True

    def clear_completed(self):
        before = len(self.tasks)
        self.tasks = [t for t in self.tasks if not t.isDone]
        if len(self.tasks) == before:
            return
        self.save()
        self.emit("changed")

    @property
    def pending_count(self) -> int:
        return sum(1 for t in self.tasks if not t.isDone)

    @property
    def done_count(self) -> int:
        return sum(1 for t in self.tasks if t.isDone)

File: macos/Sources/Frontasks/Store.swift
//  Frontasks — floating, always-on-top task list for macOS.
//  Copyright (C) 2026 Ananias Filho
//
//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU Affero General Public License as published by
//  the Free Software Foundation, either version 3 of the License, or
//  (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU Affero General Public License for more details.
//
//  You should have received a copy of the GNU Affero General Public License
//  along with this program.  If not, see <https://www.gnu.org/licenses/>.

import Foundation
import Combine

/// Uma tarefa da lista.
struct TaskItem: Identifiable, Codable, Equatable {
    var id: UUID = UUID()
    var title: String
    var isDone: Bool = false
    var createdAt: Date = .now
    var order: Int = 0
}

/// Store observável, persistida em JSON dentro de Application Support.
@MainActor
final class TaskStore: ObservableObject {
    static let shared = TaskStore()

    @Published var tasks: [TaskItem] = []

    private let url: URL
    private var saveWork: DispatchWorkItem?

    private init() {
        let dir = FileManager.default
            .urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("Frontasks", isDirectory: true)
        do {
            try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        } catch {
            NSLog("Frontasks: falha ao criar a pasta de dados — \(error).")
        }
        url = dir.appendingPathComponent("tasks.json")
        load()
    }

    // MARK: - Persistência

    func load() {
        // Primeiro uso: arquivo ainda não existe — lista vazia, sem erro.
        guard FileManager.default.fileExists(atPath: url.path) else { return }
        do {
            let data = try Data(contentsOf: url)
            let decoded = try JSONDecoder().decode([TaskItem].self, from: data)
            tasks = decoded.sorted { ($0.order, $0.createdAt) < ($1.order, $1.createdAt) }
        } catch {
            // P0: NÃO destruir o arquivo do usuário. Preserva o inválido antes que
            // qualquer save() (inclusive o flush no encerramento) o sobrescreva com [].
            backupCorruptStore(error)
        }
    }

    /// Faz backup de um `tasks.json` inválido para não perder as tarefas do usuário.
    private func backupCorruptStore(_ error: Error) {
        let fmt = DateFormatter()
        fmt.dateFormat = "yyyy-MM-dd-HHmmss"
        let backup = url.deletingLastPathComponent()
            .appendingPathComponent("tasks.corrupt-\(fmt.string(from: Date())).json")
        try? FileManager.default.copyItem(at: url, to: backup)
        NSLog("Frontasks: tasks.json inválido (\(error)). Backup preservado em \(backup.lastPathComponent). Iniciando com lista vazia.")
    }

    /// Grava imediatamente (operações pontuais e no encerramento do app).
    func save() {
        saveWork?.cancel()
        saveWork = nil
        do {
            let data = try JSONEncoder().encode(tasks)
            try data.write(to: url, options: .atomic)
        } catch {
            NSLog("Frontasks: falha ao gravar tasks.json — \(error).")
        }
    }

    /// Grava com pequeno atraso — para edição de texto (evita escrever a cada tecla).
    private func saveDebounced() {
        saveWork?.cancel()
        let work = DispatchWorkItem { [weak self] in self?.save() }
        saveWork = work
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.4, execute: work)
    }

    // MARK: - CRUD

    func add(_ title: String) {
        let trimmed = title.trimmingCharacters(in: .whitespaces)
        guard !trimmed.isEmpty else { return }
        let maxOrder = tasks.map(\.order).max() ?? 0
        tasks.append(TaskItem(title: trimmed, order: maxOrder + 1))
        save()
    }

    func toggle(_ id: UUID) {
        guard let i = tasks.firstIndex(where: { $0.id == id }) else { return }
        tasks[i].isDone.toggle()
        save()
    }

    func updateTitle(_ id: UUID, _ title: String) {
        guard let i = tasks.firstIndex(where: { $0.id == id }) else { return }
        tasks[i].title = title
        saveDebounced()
    }

    /// Finaliza a edição de um título: remove espaços das pontas e, se ficar vazio,
    /// apaga a tarefa. Deve ser chamado no commit (Enter/perda de foco), NUNCA a cada tecla.
    func commitTitle(_ id: UUID) {
        guard let i = tasks.firstIndex(where: { $0.id == id }) else { return }
        let trimmed = tasks[i].title.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty {
            tasks.remove(at: i)
        } else {
            tasks[i].title = trimmed
        }
        save()
    }

    func delete(_ id: UUID) {
        tasks.removeAll { $0.id == id }
        save()
    }

    /// Reordena (arrastar) e reatribui `order` para persistir a nova sequência.
    func move(from source: IndexSet, to destination: Int) {
        tasks.move(fromOffsets: source, toOffset: destination)
        for i in tasks.indices { tasks[i].order = i }
        save()
    }

    /// Remove de uma vez todas as tarefas concluídas.
    func clearCompleted() {
        tasks.removeAll { $0.isDone }
        save()
    }

    var pendingCount: Int { tasks.filter { !$0.isDone }.count }
    var doneCount: Int { tasks.filter { $0.isDone }.count }
}
